replica: Use flat_hash_map for tablet storage

raphaelsc opened the pull request:
The reason that we want to switch to flat_hash_map is that only a small subset of tablets will be allocated on any given shard, therefore it's wasteful to use a sparse array, and iterations are slow. Also, the map gives greater development flexibility as one doesn't have to worry about empty entries.

perf result:

```
-- reads

scylla_with_chunked_vector-read-no-tablets.txt
median 3185.23 tps (143.9 allocs/op,  41.7 tasks/op, 1376280 insns/op,        0 errors)
median 3188.02 tps (143.9 allocs/op,  41.6 tasks/op, 1375281 insns/op,        0 errors)
median 3174.20 tps (143.9 allocs/op,  41.7 tasks/op, 1376079 insns/op,        0 errors)
median 3176.89 tps (143.9 allocs/op,  41.6 tasks/op, 1376365 insns/op,        0 errors)
median 3180.71 tps (144.0 allocs/op,  41.7 tasks/op, 1375830 insns/op,        0 errors)

scylla_with_hash_map-read-no-tablets.txt
median 3173.52 tps (143.9 allocs/op,  41.6 tasks/op, 1376780 insns/op,        0 errors)
median 3160.90 tps (144.0 allocs/op,  41.7 tasks/op, 1375791 insns/op,        0 errors)
median 3184.74 tps (143.9 allocs/op,  41.6 tasks/op, 1375739 insns/op,        0 errors)
median 3162.45 tps (143.9 allocs/op,  41.7 tasks/op, 1376766 insns/op,        0 errors)
median 3160.08 tps (143.9 allocs/op,  41.7 tasks/op, 1375774 insns/op,        0 errors)

scylla_with_chunked_vector-read-tablets-1.txt
median 3168.02 tps (143.9 allocs/op,  41.7 tasks/op, 1377753 insns/op,        0 errors)
median 3159.93 tps (143.9 allocs/op,  41.6 tasks/op, 1376816 insns/op,        0 errors)
median 3174.50 tps (143.9 allocs/op,  41.7 tasks/op, 1376928 insns/op,        0 errors)
median 3173.73 tps (143.9 allocs/op,  41.7 tasks/op, 1377247 insns/op,        0 errors)
median 3170.43 tps (143.9 allocs/op,  41.6 tasks/op, 1376206 insns/op,        0 errors)

scylla_with_hash_map-read-tablets-1.txt
median 3167.78 tps (143.9 allocs/op,  41.7 tasks/op, 1376271 insns/op,        0 errors)
median 3163.32 tps (143.9 allocs/op,  41.7 tasks/op, 1377433 insns/op,        0 errors)
median 3160.78 tps (143.9 allocs/op,  41.7 tasks/op, 1375723 insns/op,        0 errors)
median 3162.10 tps (143.9 allocs/op,  41.7 tasks/op, 1376536 insns/op,        0 errors)
median 3146.68 tps (143.9 allocs/op,  41.6 tasks/op, 1378097 insns/op,        0 errors)

scylla_with_chunked_vector-read-tablets-128.txt
median 13392.89 tps (143.7 allocs/op,  41.3 tasks/op,  264531 insns/op,        0 errors)
median 13360.79 tps (143.7 allocs/op,  41.3 tasks/op,  264401 insns/op,        0 errors)
median 13330.00 tps (143.7 allocs/op,  41.3 tasks/op,  264645 insns/op,        0 errors)
median 13330.75 tps (143.7 allocs/op,  41.3 tasks/op,  264510 insns/op,        0 errors)
median 13369.72 tps (143.7 allocs/op,  41.3 tasks/op,  264439 insns/op,        0 errors)

scylla_with_hash_map-read-tablets-128.txt
median 13336.14 tps (143.7 allocs/op,  41.3 tasks/op,  264527 insns/op,        0 errors)
median 13316.85 tps (143.7 allocs/op,  41.3 tasks/op,  264661 insns/op,        0 errors)
median 13299.45 tps (143.7 allocs/op,  41.3 tasks/op,  264236 insns/op,        0 errors)
median 13311.89 tps (143.7 allocs/op,  41.3 tasks/op,  264376 insns/op,        0 errors)
median 13330.37 tps (143.7 allocs/op,  41.3 tasks/op,  264541 insns/op,        0 errors)

[1] with 1 or 128 tablets, flat_hash_map provides similar perf as chunked_vector.
```

```
-- writes

scylla_with_chunked_vector-write-no-tablets.txt
median 68635.17 tps ( 58.4 allocs/op,  13.3 tasks/op,   52709 insns/op,        0 errors)
median 68716.36 tps ( 58.4 allocs/op,  13.3 tasks/op,   52691 insns/op,        0 errors)
median 68512.76 tps ( 58.4 allocs/op,  13.3 tasks/op,   52721 insns/op,        0 errors)
median 68606.14 tps ( 58.4 allocs/op,  13.3 tasks/op,   52696 insns/op,        0 errors)
median 68619.25 tps ( 58.4 allocs/op,  13.3 tasks/op,   52697 insns/op,        0 errors)

scylla_with_hash_map-write-no-tablets.txt
median 67678.10 tps ( 58.4 allocs/op,  13.3 tasks/op,   52723 insns/op,        0 errors)
median 67966.06 tps ( 58.4 allocs/op,  13.3 tasks/op,   52736 insns/op,        0 errors)
median 67881.47 tps ( 58.4 allocs/op,  13.3 tasks/op,   52743 insns/op,        0 errors)
median 67856.81 tps ( 58.4 allocs/op,  13.3 tasks/op,   52730 insns/op,        0 errors)
median 67812.58 tps ( 58.4 allocs/op,  13.3 tasks/op,   52740 insns/op,        0 errors)


scylla_with_chunked_vector-write-tablets-1.txt
median 67741.83 tps ( 58.4 allocs/op,  13.3 tasks/op,   53425 insns/op,        0 errors)
median 68014.20 tps ( 58.4 allocs/op,  13.3 tasks/op,   53455 insns/op,        0 errors)
median 68228.48 tps ( 58.4 allocs/op,  13.3 tasks/op,   53447 insns/op,        0 errors)
median 67950.96 tps ( 58.4 allocs/op,  13.3 tasks/op,   53443 insns/op,        0 errors)
median 67832.69 tps ( 58.4 allocs/op,  13.3 tasks/op,   53462 insns/op,        0 errors)

scylla_with_hash_map-write-tablets-1.txt
median 66873.70 tps ( 58.4 allocs/op,  13.3 tasks/op,   53548 insns/op,        0 errors)
median 67568.23 tps ( 58.4 allocs/op,  13.3 tasks/op,   53547 insns/op,        0 errors)
median 67653.70 tps ( 58.4 allocs/op,  13.3 tasks/op,   53525 insns/op,        0 errors)
median 67389.21 tps ( 58.4 allocs/op,  13.3 tasks/op,   53536 insns/op,        0 errors)
median 67437.91 tps ( 58.4 allocs/op,  13.3 tasks/op,   53537 insns/op,        0 errors)


scylla_with_chunked_vector-write-tablets-128.txt
median 67115.41 tps ( 58.3 allocs/op,  13.3 tasks/op,   53341 insns/op,        0 errors)
median 66836.07 tps ( 58.3 allocs/op,  13.3 tasks/op,   53342 insns/op,        0 errors)
median 67214.07 tps ( 58.3 allocs/op,  13.3 tasks/op,   53303 insns/op,        0 errors)
median 67198.25 tps ( 58.3 allocs/op,  13.3 tasks/op,   53347 insns/op,        0 errors)
median 67368.78 tps ( 58.3 allocs/op,  13.3 tasks/op,   53374 insns/op,        0 errors)

scylla_with_hash_map-write-tablets-128.txt
median 66273.50 tps ( 58.3 allocs/op,  13.3 tasks/op,   53400 insns/op,        0 errors)
median 66564.89 tps ( 58.3 allocs/op,  13.3 tasks/op,   53432 insns/op,        0 errors)
median 66568.52 tps ( 58.3 allocs/op,  13.3 tasks/op,   53408 insns/op,        0 errors)
median 66368.00 tps ( 58.3 allocs/op,  13.3 tasks/op,   53441 insns/op,        0 errors)
median 66293.55 tps ( 58.3 allocs/op,  13.3 tasks/op,   53408 insns/op,        0 errors)
```

Fixes #18010.

raphaelsc commented:
this is the script used to generate the results: https://gist.githubusercontent.com/raphaelsc/2c9eaaedc1947a950f4ea19ca7966d9c/raw/fc99e52e7be4587d855e9b254485660132346651/bench.sh

cpu frequency was also set, to reduce noise.

avikivity commented:
I don't understand these results. 137k ops/read? what workload is this?

Typical is ~50k IIRC.

avikivity commented on absl-flat_hash_map.hh:
```diff
@@ -29,6 +38,10 @@ template <typename K, typename V, typename... Ts>
+template <typename V>
+struct flat_hash_map<size_t, V>
+        : public absl::flat_hash_map<size_t, V, integral_hash<size_t>> {};
+
```
I think we can use absl::Hash and absl::flat_hash_map directly. IIRC absl-flat_hash_map.hh was introduced to work around some integration problems with sstring, but here we can just use it directly.

Especially since the new flat_hash_map eats the key type parameter.

raphaelsc replied:
Acked.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7785/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7785/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/751/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8778/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7785/)
#### Build Details:
- Duration: 2 hr 33 min
- Builder: spider7.cloudius-systems.com

raphaelsc commented:
> I don't understand these results. 137k ops/read? what workload is this?
> 
> Typical is ~50k IIRC.

bypassed cache, intentionally consuming from sstables to stress tablet_sstable_set. note index caching is also bypassed. workload is random reads by default, so we may be rereading the same index pages more than once, yielding high read ampl.

raphaelsc commented:
@bhalevy please review the changes in tablet_sstable_set

bhalevy commented on replica/table.cc:
```diff
@@ -559,10 +560,11 @@ class single_storage_group_manager final : public storage_group_manager {
-        r.push_back(std::make_unique<storage_group>(std::move(cg), &_compaction_groups));
+        r[0] = std::make_unique<storage_group>(std::move(cg), &_compaction_groups);
         _storage_groups = std::move(r);
+        _single_sg = _storage_groups.at(0).get();
```
could simplify this to cg.get() above, before it's moved into the map.

bhalevy commented on replica/table.cc:
```diff
@@ -898,7 +899,7 @@ utils::chunked_vector<compaction_group*> tablet_storage_group_manager::compactio
 
     while (candidate_start <= candidate_end) {
-        auto& sg = storage_groups()[candidate_start++];
+        auto& sg = storage_groups().at(candidate_start++);
```
This won't work with dynamic allocation of storage groups
since `at()` would throw if the sg isn't populated.

bhalevy commented on replica/table.cc:
```diff
@@ -2010,7 +2011,7 @@ locator::table_load_stats table::table_load_stats(std::function<bool(locator::gl
 
     for (unsigned id = 0; id < storage_groups().size(); id++) {
-        auto& sg = storage_groups()[id];
+        auto& sg = storage_groups().at(id);
```
ditto, `at(id)` would throw.
wouldn't `for (const auto& [id, sg] : storage_groups())` work better here?
and then the (!sg) below would probably need to be an internal error after dynamic allocation of storage groups is implemented.

raphaelsc replied:
yes, it should have been "for (const auto& [id, sg] : storage_groups())", thinking forward to dynamic allocation. thanks for catching it!

avikivity commented:
> > I don't understand these results. 137k ops/read? what workload is this?
> > Typical is ~50k IIRC.
> 
> bypassed cache, intentionally consuming from sstables to stress tablet_sstable_set. note index caching is also bypassed. workload is random reads by default, so we may be rereading the same index pages more than once, yielding high read ampl.

Ok, please mention it in the commit, to avoid giving people heart attacks.

avikivity commented:
I don't understand writes either. 80,000 insns/op?

avikivity commented:
btw, even cached reads should show the difference, since we read from memtables.

avikivity commented:
So it's better to use cached reads. As it is, the sstable overhead drowns what we're trying to measure.

raphaelsc commented:
> btw, even cached reads should show the difference, since we read from memtables.

memtables are flushed with --flush flag.

for write, I used:
`        ./build/dev/$SCYLLA_BINARY_NAME perf-simple-query --smp 1 -m 1G --operations-per-shard=100000 --write --tablets --initial-tablets $INITIAL_TABLETS`
for read...:
`        ./build/dev/$SCYLLA_BINARY_NAME perf-simple-query --smp 1 -m 1G --operations-per-shard=10000 --flush --bypass-cache --tablets --initial-tablets $INITIAL_TABLETS`

raphaelsc commented:
> So it's better to use cached reads. As it is, the sstable overhead drowns what we're trying to measure.

the reason I decided for bypassing cache is that cached reads won't show the difference caused by this change, since row cache is shared between groups and we don't have to access any storage group (through the data structure) when consuming from it. uncached reads on the other hand will do it. maybe write test is enough, since we'll access storage group to look up the memtable to append to.

raphaelsc commented:
> I don't understand writes either. 80,000 insns/op?

what's the common value? master shows me 79.4k with write workload. maybe a regression at some point?

avikivity commented:
> > So it's better to use cached reads. As it is, the sstable overhead drowns what we're trying to measure.
> 
> the reason I decided for bypassing cache is that cached reads won't show the difference caused by this change, since row cache is shared between groups and we don't have to access any storage group (through the data structure) when consuming from it. uncached reads on the other hand will do it. maybe write test is enough, since we'll access storage group to look up the memtable to append to.

Even if the memtables are empty, we still have to look them up. So read from cache/memtable is fine.

As it is, looking for a 100 insn regression in a 1.3M workload isn't possible.

avikivity commented:
> > I don't understand writes either. 80,000 insns/op?
> 
> what's the common value? master shows me 79.4k with write workload. maybe a regression at some point?

Maybe there was.  I remember much smaller values.

raphaelsc commented:
> > I don't understand writes either. 80,000 insns/op?
> 
> what's the common value? master shows me 79.4k with write workload. maybe a regression at some point?

oops, I think the problem may be that I was doing the perf test in dev mode. Sorry.

raphaelsc commented:
> > > So it's better to use cached reads. As it is, the sstable overhead drowns what we're trying to measure.
> > 
> > 
> > the reason I decided for bypassing cache is that cached reads won't show the difference caused by this change, since row cache is shared between groups and we don't have to access any storage group (through the data structure) when consuming from it. uncached reads on the other hand will do it. maybe write test is enough, since we'll access storage group to look up the memtable to append to.
> 
> Even if the memtables are empty, we still have to look them up. So read from cache/memtable is fine.
> 
> As it is, looking for a 100 insn regression in a 1.3M workload isn't possible.

Good points, will redo the tests.

avikivity commented:
> > > I don't understand writes either. 80,000 insns/op?
> > 
> > 
> > what's the common value? master shows me 79.4k with write workload. maybe a regression at some point?
> 
> oops, I think the problem may be that I was doing the perf test in dev mode. Sorry.

Ok be careful I can't take many more of these perf-simple-query regressions.

raphaelsc commented:
new results:

```
-- reads

scylla_with_chunked_vector-read-no-tablets.txt
median 73223.28 tps ( 62.3 allocs/op,  13.3 tasks/op,   41932 insns/op,        0 errors)
median 74952.87 tps ( 62.3 allocs/op,  13.3 tasks/op,   41969 insns/op,        0 errors)
median 73016.37 tps ( 62.3 allocs/op,  13.3 tasks/op,   41934 insns/op,        0 errors)
median 74078.14 tps ( 62.3 allocs/op,  13.3 tasks/op,   41938 insns/op,        0 errors)
median 75323.07 tps ( 62.3 allocs/op,  13.3 tasks/op,   41944 insns/op,        0 errors)

scylla_with_hash_map-read-no-tablets.txt
median 74963.30 tps ( 62.3 allocs/op,  13.3 tasks/op,   41926 insns/op,        0 errors)
median 74032.09 tps ( 62.3 allocs/op,  13.3 tasks/op,   41918 insns/op,        0 errors)
median 74850.09 tps ( 62.3 allocs/op,  13.3 tasks/op,   41937 insns/op,        0 errors)
median 74239.37 tps ( 62.3 allocs/op,  13.3 tasks/op,   41921 insns/op,        0 errors)
median 74798.14 tps ( 62.3 allocs/op,  13.3 tasks/op,   41925 insns/op,        0 errors)

scylla_with_chunked_vector-read-tablets-1.txt
median 74234.27 tps ( 62.1 allocs/op,  13.3 tasks/op,   41903 insns/op,        0 errors)
median 75775.98 tps ( 62.1 allocs/op,  13.3 tasks/op,   41910 insns/op,        0 errors)
median 76481.56 tps ( 62.1 allocs/op,  13.2 tasks/op,   41874 insns/op,        0 errors)
median 74056.67 tps ( 62.1 allocs/op,  13.3 tasks/op,   41894 insns/op,        0 errors)
median 75287.68 tps ( 62.1 allocs/op,  13.3 tasks/op,   41894 insns/op,        0 errors)

scylla_with_hash_map-read-tablets-1.txt
median 75613.63 tps ( 62.1 allocs/op,  13.2 tasks/op,   41990 insns/op,        0 errors)
median 74819.51 tps ( 62.1 allocs/op,  13.2 tasks/op,   41973 insns/op,        0 errors)
median 75648.41 tps ( 62.1 allocs/op,  13.3 tasks/op,   42025 insns/op,        0 errors)
median 74170.89 tps ( 62.1 allocs/op,  13.2 tasks/op,   42002 insns/op,        0 errors)
median 75447.72 tps ( 62.1 allocs/op,  13.3 tasks/op,   41952 insns/op,        0 errors)


scylla_with_chunked_vector-read-tablets-128.txt
median 73788.57 tps ( 62.1 allocs/op,  13.2 tasks/op,   41956 insns/op,        0 errors)
median 76563.63 tps ( 62.1 allocs/op,  13.3 tasks/op,   42006 insns/op,        0 errors)
median 75536.12 tps ( 62.1 allocs/op,  13.2 tasks/op,   42005 insns/op,        0 errors)
median 74679.17 tps ( 62.1 allocs/op,  13.3 tasks/op,   41958 insns/op,        0 errors)
median 75380.95 tps ( 62.1 allocs/op,  13.2 tasks/op,   41946 insns/op,        0 errors)

scylla_with_hash_map-read-tablets-128.txt
median 75459.99 tps ( 62.1 allocs/op,  13.3 tasks/op,   42055 insns/op,        0 errors)
median 74280.11 tps ( 62.1 allocs/op,  13.3 tasks/op,   42085 insns/op,        0 errors)
median 74502.61 tps ( 62.1 allocs/op,  13.3 tasks/op,   42063 insns/op,        0 errors)
median 74692.27 tps ( 62.1 allocs/op,  13.3 tasks/op,   41994 insns/op,        0 errors)
median 75402.64 tps ( 62.1 allocs/op,  13.3 tasks/op,   42015 insns/op,        0 errors)

```

```
-- writes

scylla_with_chunked_vector-write-no-tablets.txt
median 68635.17 tps ( 58.4 allocs/op,  13.3 tasks/op,   52709 insns/op,        0 errors)
median 68716.36 tps ( 58.4 allocs/op,  13.3 tasks/op,   52691 insns/op,        0 errors)
median 68512.76 tps ( 58.4 allocs/op,  13.3 tasks/op,   52721 insns/op,        0 errors)
median 68606.14 tps ( 58.4 allocs/op,  13.3 tasks/op,   52696 insns/op,        0 errors)
median 68619.25 tps ( 58.4 allocs/op,  13.3 tasks/op,   52697 insns/op,        0 errors)

scylla_with_hash_map-write-no-tablets.txt
median 67678.10 tps ( 58.4 allocs/op,  13.3 tasks/op,   52723 insns/op,        0 errors)
median 67966.06 tps ( 58.4 allocs/op,  13.3 tasks/op,   52736 insns/op,        0 errors)
median 67881.47 tps ( 58.4 allocs/op,  13.3 tasks/op,   52743 insns/op,        0 errors)
median 67856.81 tps ( 58.4 allocs/op,  13.3 tasks/op,   52730 insns/op,        0 errors)
median 67812.58 tps ( 58.4 allocs/op,  13.3 tasks/op,   52740 insns/op,        0 errors)


scylla_with_chunked_vector-write-tablets-1.txt
median 67741.83 tps ( 58.4 allocs/op,  13.3 tasks/op,   53425 insns/op,        0 errors)
median 68014.20 tps ( 58.4 allocs/op,  13.3 tasks/op,   53455 insns/op,        0 errors)
median 68228.48 tps ( 58.4 allocs/op,  13.3 tasks/op,   53447 insns/op,        0 errors)
median 67950.96 tps ( 58.4 allocs/op,  13.3 tasks/op,   53443 insns/op,        0 errors)
median 67832.69 tps ( 58.4 allocs/op,  13.3 tasks/op,   53462 insns/op,        0 errors)

scylla_with_hash_map-write-tablets-1.txt
median 66873.70 tps ( 58.4 allocs/op,  13.3 tasks/op,   53548 insns/op,        0 errors)
median 67568.23 tps ( 58.4 allocs/op,  13.3 tasks/op,   53547 insns/op,        0 errors)
median 67653.70 tps ( 58.4 allocs/op,  13.3 tasks/op,   53525 insns/op,        0 errors)
median 67389.21 tps ( 58.4 allocs/op,  13.3 tasks/op,   53536 insns/op,        0 errors)
median 67437.91 tps ( 58.4 allocs/op,  13.3 tasks/op,   53537 insns/op,        0 errors)


scylla_with_chunked_vector-write-tablets-128.txt
median 67115.41 tps ( 58.3 allocs/op,  13.3 tasks/op,   53341 insns/op,        0 errors)
median 66836.07 tps ( 58.3 allocs/op,  13.3 tasks/op,   53342 insns/op,        0 errors)
median 67214.07 tps ( 58.3 allocs/op,  13.3 tasks/op,   53303 insns/op,        0 errors)
median 67198.25 tps ( 58.3 allocs/op,  13.3 tasks/op,   53347 insns/op,        0 errors)
median 67368.78 tps ( 58.3 allocs/op,  13.3 tasks/op,   53374 insns/op,        0 errors)

scylla_with_hash_map-write-tablets-128.txt
median 66273.50 tps ( 58.3 allocs/op,  13.3 tasks/op,   53400 insns/op,        0 errors)
median 66564.89 tps ( 58.3 allocs/op,  13.3 tasks/op,   53432 insns/op,        0 errors)
median 66568.52 tps ( 58.3 allocs/op,  13.3 tasks/op,   53408 insns/op,        0 errors)
median 66368.00 tps ( 58.3 allocs/op,  13.3 tasks/op,   53441 insns/op,        0 errors)
median 66293.55 tps ( 58.3 allocs/op,  13.3 tasks/op,   53408 insns/op,        0 errors)
```

avikivity commented:
Much better. Looks like write+tablet has a large regression, but not related to this change.

raphaelsc commented:
> new results:
> 
> ```
> -- reads
> 
> scylla_with_chunked_vector-read-no-tablets.txt
> median 73223.28 tps ( 62.3 allocs/op,  13.3 tasks/op,   41932 insns/op,        0 errors)
> median 74952.87 tps ( 62.3 allocs/op,  13.3 tasks/op,   41969 insns/op,        0 errors)
> median 73016.37 tps ( 62.3 allocs/op,  13.3 tasks/op,   41934 insns/op,        0 errors)
> median 74078.14 tps ( 62.3 allocs/op,  13.3 tasks/op,   41938 insns/op,        0 errors)
> median 75323.07 tps ( 62.3 allocs/op,  13.3 tasks/op,   41944 insns/op,        0 errors)
> 
> scylla_with_hash_map-read-no-tablets.txt
> median 74963.30 tps ( 62.3 allocs/op,  13.3 tasks/op,   41926 insns/op,        0 errors)
> median 74032.09 tps ( 62.3 allocs/op,  13.3 tasks/op,   41918 insns/op,        0 errors)
> median 74850.09 tps ( 62.3 allocs/op,  13.3 tasks/op,   41937 insns/op,        0 errors)
> median 74239.37 tps ( 62.3 allocs/op,  13.3 tasks/op,   41921 insns/op,        0 errors)
> median 74798.14 tps ( 62.3 allocs/op,  13.3 tasks/op,   41925 insns/op,        0 errors)
> 
> scylla_with_chunked_vector-read-tablets-1.txt
> median 74234.27 tps ( 62.1 allocs/op,  13.3 tasks/op,   41903 insns/op,        0 errors)
> median 75775.98 tps ( 62.1 allocs/op,  13.3 tasks/op,   41910 insns/op,        0 errors)
> median 76481.56 tps ( 62.1 allocs/op,  13.2 tasks/op,   41874 insns/op,        0 errors)
> median 74056.67 tps ( 62.1 allocs/op,  13.3 tasks/op,   41894 insns/op,        0 errors)
> median 75287.68 tps ( 62.1 allocs/op,  13.3 tasks/op,   41894 insns/op,        0 errors)
> 
> scylla_with_hash_map-read-tablets-1.txt
> median 75613.63 tps ( 62.1 allocs/op,  13.2 tasks/op,   41990 insns/op,        0 errors)
> median 74819.51 tps ( 62.1 allocs/op,  13.2 tasks/op,   41973 insns/op,        0 errors)
> median 75648.41 tps ( 62.1 allocs/op,  13.3 tasks/op,   42025 insns/op,        0 errors)
> median 74170.89 tps ( 62.1 allocs/op,  13.2 tasks/op,   42002 insns/op,        0 errors)
> median 75447.72 tps ( 62.1 allocs/op,  13.3 tasks/op,   41952 insns/op,        0 errors)
> 
> 
> scylla_with_chunked_vector-read-tablets-128.txt
> median 73788.57 tps ( 62.1 allocs/op,  13.2 tasks/op,   41956 insns/op,        0 errors)
> median 76563.63 tps ( 62.1 allocs/op,  13.3 tasks/op,   42006 insns/op,        0 errors)
> median 75536.12 tps ( 62.1 allocs/op,  13.2 tasks/op,   42005 insns/op,        0 errors)
> median 74679.17 tps ( 62.1 allocs/op,  13.3 tasks/op,   41958 insns/op,        0 errors)
> median 75380.95 tps ( 62.1 allocs/op,  13.2 tasks/op,   41946 insns/op,        0 errors)
> 
> scylla_with_hash_map-read-tablets-128.txt
> median 75459.99 tps ( 62.1 allocs/op,  13.3 tasks/op,   42055 insns/op,        0 errors)
> median 74280.11 tps ( 62.1 allocs/op,  13.3 tasks/op,   42085 insns/op,        0 errors)
> median 74502.61 tps ( 62.1 allocs/op,  13.3 tasks/op,   42063 insns/op,        0 errors)
> median 74692.27 tps ( 62.1 allocs/op,  13.3 tasks/op,   41994 insns/op,        0 errors)
> median 75402.64 tps ( 62.1 allocs/op,  13.3 tasks/op,   42015 insns/op,        0 errors)
> ```
> 
> ```
> -- writes
> 
> scylla_with_chunked_vector-write-no-tablets.txt
> median 68635.17 tps ( 58.4 allocs/op,  13.3 tasks/op,   52709 insns/op,        0 errors)
> median 68716.36 tps ( 58.4 allocs/op,  13.3 tasks/op,   52691 insns/op,        0 errors)
> median 68512.76 tps ( 58.4 allocs/op,  13.3 tasks/op,   52721 insns/op,        0 errors)
> median 68606.14 tps ( 58.4 allocs/op,  13.3 tasks/op,   52696 insns/op,        0 errors)
> median 68619.25 tps ( 58.4 allocs/op,  13.3 tasks/op,   52697 insns/op,        0 errors)
> 
> scylla_with_hash_map-write-no-tablets.txt
> median 67678.10 tps ( 58.4 allocs/op,  13.3 tasks/op,   52723 insns/op,        0 errors)
> median 67966.06 tps ( 58.4 allocs/op,  13.3 tasks/op,   52736 insns/op,        0 errors)
> median 67881.47 tps ( 58.4 allocs/op,  13.3 tasks/op,   52743 insns/op,        0 errors)
> median 67856.81 tps ( 58.4 allocs/op,  13.3 tasks/op,   52730 insns/op,        0 errors)
> median 67812.58 tps ( 58.4 allocs/op,  13.3 tasks/op,   52740 insns/op,        0 errors)
> 
> 
> scylla_with_chunked_vector-write-tablets-1.txt
> median 67741.83 tps ( 58.4 allocs/op,  13.3 tasks/op,   53425 insns/op,        0 errors)
> median 68014.20 tps ( 58.4 allocs/op,  13.3 tasks/op,   53455 insns/op,        0 errors)
> median 68228.48 tps ( 58.4 allocs/op,  13.3 tasks/op,   53447 insns/op,        0 errors)
> median 67950.96 tps ( 58.4 allocs/op,  13.3 tasks/op,   53443 insns/op,        0 errors)
> median 67832.69 tps ( 58.4 allocs/op,  13.3 tasks/op,   53462 insns/op,        0 errors)
> 
> scylla_with_hash_map-write-tablets-1.txt
> median 66873.70 tps ( 58.4 allocs/op,  13.3 tasks/op,   53548 insns/op,        0 errors)
> median 67568.23 tps ( 58.4 allocs/op,  13.3 tasks/op,   53547 insns/op,        0 errors)
> median 67653.70 tps ( 58.4 allocs/op,  13.3 tasks/op,   53525 insns/op,        0 errors)
> median 67389.21 tps ( 58.4 allocs/op,  13.3 tasks/op,   53536 insns/op,        0 errors)
> median 67437.91 tps ( 58.4 allocs/op,  13.3 tasks/op,   53537 insns/op,        0 errors)
> 
> 
> scylla_with_chunked_vector-write-tablets-128.txt
> median 67115.41 tps ( 58.3 allocs/op,  13.3 tasks/op,   53341 insns/op,        0 errors)
> median 66836.07 tps ( 58.3 allocs/op,  13.3 tasks/op,   53342 insns/op,        0 errors)
> median 67214.07 tps ( 58.3 allocs/op,  13.3 tasks/op,   53303 insns/op,        0 errors)
> median 67198.25 tps ( 58.3 allocs/op,  13.3 tasks/op,   53347 insns/op,        0 errors)
> median 67368.78 tps ( 58.3 allocs/op,  13.3 tasks/op,   53374 insns/op,        0 errors)
> 
> scylla_with_hash_map-write-tablets-128.txt
> median 66273.50 tps ( 58.3 allocs/op,  13.3 tasks/op,   53400 insns/op,        0 errors)
> median 66564.89 tps ( 58.3 allocs/op,  13.3 tasks/op,   53432 insns/op,        0 errors)
> median 66568.52 tps ( 58.3 allocs/op,  13.3 tasks/op,   53408 insns/op,        0 errors)
> median 66368.00 tps ( 58.3 allocs/op,  13.3 tasks/op,   53441 insns/op,        0 errors)
> median 66293.55 tps ( 58.3 allocs/op,  13.3 tasks/op,   53408 insns/op,        0 errors)
> ```

with 128 tablets, read and write have lost in tps 0.5 and 1% respectively. maybe some of that is due to noise, but I made sure to set fixed cpu frequency.

raphaelsc commented:
> Much better. Looks like write+tablet has a large regression, but not related to this change.

perf diff allows us to inspect calls that are only present with tablets

we have for example compaction_group_for_key() which has to calculate token again from partition key, and that doesn't exist in vnode mode. maybe some of that could also come from tablet sharder, which has to lookup tablet map on every write I think.

```
# Event 'cycles:Pu'
#
# Baseline  Delta Abs  Shared Object                       Symbol                                                                                                                                                                                                                                                            >
# ........  .........  ..................................  ..................................................................................................................................................................................................................................................................>
#
               +0.51%  scylla                              [.] _ZNK7replica28tablet_storage_group_manager24compaction_group_for_keyE18partition_key_viewRKN7seastar13lw_shared_ptrIK6schemaEE
               +0.44%  scylla                              [.] _ZNK7locator32tablet_effective_replication_map21get_natural_endpointsERKN3dht5tokenE
     1.00%     -0.43%  scylla                              [.] _ZN3dhtssERKNS_5tokenES2_
     0.59%     +0.41%  scylla                              [.] _ZN7replica8memtable5applyERK15frozen_mutationRKN7seastar13lw_shared_ptrIK6schemaEEON2db9rp_handleE
               +0.37%  scylla                              [.] _ZNK7locator32tablet_effective_replication_map14check_localityERKN3dht5tokenE
     1.47%     +0.34%  scylla                              [.] _ZNK3dht19murmur3_partitioner9get_tokenERK6schema18partition_key_view
     2.22%     -0.28%  scylla                              [.] _ZN18compact_radix_tree4treeI13cell_and_hashjE10visit_slotIRNS2_15walking_visitorIZNK3row13for_each_cellIZN12_GLOBAL__N_115write_row_cellsIN3ser20deletable_row__cellsI13bytes_ostreamEEEEDaOT_RKS5_RK6schema11column_kindEUljRK25atomic_cell_or_collectionE_E>
     1.93%     +0.28%  [vdso]                              [.] __vdso_clock_gettime
     2.15%     +0.26%  scylla                              [.] _ZN13compound_typeIL14allow_prefixes0EE8iteratorC2ERK24managed_bytes_basic_viewIL12mutable_view0EE
     1.54%     -0.24%  libc.so.6                           [.] __memmove_avx_unaligned_erms
     0.61%     +0.24%  [unknown]                           [k] 0xffffffff87044240
     2.06%     -0.23%  scylla                              [.] _ZN8logalloc11region_impl4freeEPvm
     1.52%     -0.21%  scylla                              [.] _ZN21mutation_partition_v219apply_monotonicallyERK6schemaS2_OS_P13cache_trackerR26mutation_application_statsN7seastar20noncopyable_functionIDoFbvEEER12apply_resumeNS8_10bool_classI13evictable_tagEE
     0.72%     -0.19%  scylla                              [.] _ZN7service13storage_proxy12mutate_beginEN5utils12small_vectorINS0_23unique_response_handlerELm1EEEN2db17consistency_levelEN7tracing15trace_state_ptrESt8optionalINSt6chrono10time_pointIN7seastar12lowres_clockENSA_8durationIlSt5ratioILl1ELl1000000000EEEEE>
     0.81%     -0.19%  scylla                              [.] _ZN7service13storage_proxy9do_mutateESt6vectorI8mutationSaIS2_EEN2db17consistency_levelENSt6chrono10time_pointIN7seastar12lowres_clockENS7_8durationIlSt5ratioILl1ELl1000000000EEEEEEN7tracing15trace_state_ptrE14service_permitbNS9_10bool_classINS5_34allow_>
     1.01%     -0.18%  scylla                              [.] _ZN7seastar7reactor14run_some_tasksEv
               +0.18%  scylla                              [.] _ZNK7locator14tablet_sharder8shard_ofERKN3dht5tokenE
     0.25%     +0.16%  scylla                              [.] _ZN15partition_entry5applyERN8logalloc6regionER16mutation_cleanerRK6schemaO21mutation_partition_v2S7_R26mutation_application_stats
     0.81%     -0.14%  scylla                              [.] _ZN8logalloc11region_impl5allocEPK15migrate_fn_typemm
     2.83%     -0.14%  scylla                              [.] _ZN13compound_typeIL14allow_prefixes1EE8iteratorC2ERK24managed_bytes_basic_viewIL12mutable_view0EE
     0.58%     -0.14%  scylla                              [.] _ZN7seastar6memory10small_pool16add_more_objectsEv
     0.51%     +0.13%  scylla                              [.] _ZNK18compact_radix_tree4treeI13cell_and_hashjE9node_head5cloneIRZN3rowC1ERK6schema11column_kindRKS5_E3$_0EESt4pairIPS3_NSt15_
```

we can also for example tag first condition as likely when getting write replicas, as tablet in transition is the unlikely scenario:

```
        auto&& replicas = std::invoke([&] () -> const tablet_replica_set& {
            if (!info) {
                return tablets.get_tablet_info(tablet).replicas;
            }
```

avikivity commented:
75k tps -> 13 usec/op.

1% of that = 130ns

With many tablets, the metadata is much larger, and so we'll get cache misses. It can translate to one cache miss all the way to memory (unlikely) or several cache misses to L3.

With the vector, the CPU just ignores the cache misses and moves on, and picks up the memory load later. With the hash table, it can't ignore the cache misses since it needs to make unpredictable decisions on the results.

We just have to live with it.

raphaelsc commented:
v2:
- use absl::flat_hash_map and absl::Hash directly 
- adjusted tablet_storage_group_manager::compaction_groups_for_token_range() and table::table_load_stats()
- updated commit msg with new results

avikivity commented on replica/tablets.cc:
```diff
@@ -474,10 +479,8 @@ stop_iteration tablet_sstable_set::for_each_sstable_set_until(const dht::partiti
-    for (auto i = candidate_start; i <= candidate_end; i++) {
-        if (const auto& set = _sstable_sets[i]) {
+    for (const auto& i : subrange(pr)) {
+        if (const auto& set = _sstable_sets.at(i)) {
```
We shouldn't need to iterate over more than one tablet here. But let's leave that until multishard_reader dies.

avikivity replied:
@denesb @bhalevy can this be called with a tablet-spanning range?

avikivity replied:
storage_proxy::query_partition_key_range() needs to be taught about tablets:

1. hold erm
2. clamp range into first tablet that intersects it, remember if we changed the range
3. read page
4. has_more_pages |= we changed the range
5. return page

raphaelsc replied:
we also have 

```
std::vector<sstables::shared_sstable> table::select_sstables(const dht::partition_range& range) const {
    return _sstables->select(range);
}
```

used by mutation dumper. we need to inspect every call in replica::table that is supplied with a token range, which can potentially cross tablets

if we come to think of it, tablet sstable set was introduced such that it can provide correct behavior while crossing tablets, isn't it good that we keep this capability even if we patch users to receive tablet treatment? It's orthogonal, and if it turns out one area or another can still cross tablets, we get correct behavior rather than aborting.

avikivity replied:
There's never a reason to cross tablet replica bounds locally. If some shard owns tablet replica 37 and tablet replica 89, why would we want to read 89 just after 37?

Even SELECT mutation_fragments needs to jump among shards. So it should read shard 52 on a neighboring shard before 89.

Think of tablet replicas as independent nodes in a mini-rack. Would you want to read data from all nodes in a mini-rack? There's no relationship between them other than their physical location.

What we have here is baggage from multishard_reader, which in turn is due to tiny data ranges that made jumping between shards too expensive to do sequentially, so we did it concurrently. With tablets there's no such problem. It's more correct for SELECT mutation_fragment to chop ranges into tablets itself, or even better ask storage_proxy to do it.

avikivity commented:
Looks good. @bhalevy please review the tablet_sstable_set changes.

bhalevy commented on replica/table.cc:
```diff
@@ -666,11 +672,11 @@ class tablet_storage_group_manager final : public storage_group_manager {
+                                                    idx, log2_storage_groups(), tablet_count(), t));
         }
-        auto& sg = storage_groups()[idx];
+        auto& sg = storage_groups().at(idx);
```
this is not backward compatible. I'm not sure if this function can be called for non-existent storage groups or not, but if it can, we should either catch and ignore the out_of_range exception here, or check if storage_groups().contains(idx) first for the sanity check.

avikivity replied:
How can it be called for non-existent storage groups?

bhalevy replied:
From `storage_group_for_token`.
Even if we end up in requiring the storage_group's existence on this path, it should be done in a follow-up series, and leave this change backward compatible with the existing code, otherwise, we'll be changing too many things at once.

bhalevy replied:
This code happens to work now, but it will be fragile with dynamic allocation of storage_groups, unless we make sure we deal with non-existing storage_groups correctly.

bhalevy replied:
This function returns the index of the storage_group that holds a token, not a reference to the  storage_group.
There is no documentation saying the corresponding storage_group must be allocated on this shard.
We can change the semantics if we want to, but I still think we should do that in 2 steps so we can review this PR more easily and make it sure it doesn't break anything before applying the next step[, and second - we have to document the semantics of these functions, and/or have a self-explanatory api, like returning a storage_group& here rather than size_t.

bhalevy replied:
Only for the sanity check, and we don't know if it breaks with dynamic storage_group allocation.
@Deexie ?

raphaelsc replied:
I am adding the sanity check as you proposed, will send new version and we can continue from there.

bhalevy commented on replica/table.cc:
```diff
@@ -685,7 +691,7 @@ class tablet_storage_group_manager final : public storage_group_manager {
     }
     storage_group* storage_group_for_token(dht::token token) const noexcept override {
-        return storage_groups()[storage_group_of(token).first].get();
+        return storage_groups().at(storage_group_of(token).first).get();
```
This function is marked noexcept and should return nullptr if the storage_groups doesn't exist.
So better use find() instead of `at`

raphaelsc replied:
we shouldn't allow nullptr. that should trigger an exception. why is this function marked as noexcept?

raphaelsc replied:
we should actually abort, because we assert we'll always find storage group allocated.

bhalevy replied:
If so, then the function signature can be changed to return a `storage_group&` instead of a pointer.
And this must be documented in `storage_group_manager` api.

bhalevy commented on replica/table.cc:
```diff
@@ -864,7 +870,7 @@ compaction_group* table::single_compaction_group_if_available() const noexcept {
 
 compaction_group* table::get_compaction_group(size_t id) const noexcept {
-    return storage_groups()[id]->main_compaction_group().get();
+    return storage_groups().at(id)->main_compaction_group().get();
```
maybe more elegantly, it should check for the storage_group existence first, and call on_fatal_internal_error if it doesn't exist - this will produce a better error report than terminating due to exception in noexcept context.

bhalevy commented on replica/table.cc:
```diff
@@ -895,13 +901,14 @@ utils::chunked_vector<compaction_group*> tablet_storage_group_manager::compactio
-        auto& sg = storage_groups()[candidate_start++];
-        if (!sg) {
+        auto it = storage_groups().find(candidate_start++);
+        if (it == storage_groups().end()) {
```
Hmm, this is pretty expensive if we had an auxiliary std::set with the occupied indices like the one you added for the tablet_sstable_set we could use to it go over only the ids in the given range.

Also, this is apparently used only for `take_storage_snapshot`, right?
And that is used only for file-based streaming of a single tablet, so the match should be of a single storage_group.  Therefore, candidate_start should be incremented by one, but rather the next one should be derived from `get_first_token` of the next tablet_id (if it's not out of range) - with that we don't need to auxiliary std::set...

raphaelsc replied:
I am thinking about changing the interface to be tablet id oriented, like cleanup_tablet() so we can remove this compaction_groups_for_token_range(). file based streaming doesn't cross tablets anyway.

Maybe, for now, I can iterate through storage group instead, and use [candidate_start, candidate_end] range, to filter out. So we'd be iterating at present tablets in this shard only. It's essentially the same as finding the next tablet id, as hash map is unordered and you have to loop over elements anyway.

bhalevy replied:
But this has linear complexity.  Why not have `std::set<tablet_id>` as the one you propose for tablet_sstable_set?

raphaelsc replied:
but N is a small constant, right? 100. and I suggest that we change the function to be tablet id oriented, so we can get rid of this intersection, as it's not needed in practice.

raphaelsc replied:
Actually I think the impl in this version is actually good, until we change it to be tablet id oriented. if we assume that the range overlaps with 1 tablet only, then we'd be iterating at most through 2 elements (as first or last token can be exclusive).

bhalevy replied:
100 is not THAT small :)
if `[candidate_start, candidate_end]` are expected to be confined to a single tablet we should take advantage of that and look up only this tablet and nothing more.

bhalevy replied:
> Actually I think the impl in this version is actually good, until we change it to be tablet id oriented. if we assume that the range overlaps with 1 tablet only, then we'd be iterating at most through 2 elements (as first or last token can be exclusive).

Oh, what I had in mind is that candidate_start and candidate_end are tokens, not indices.
In that case this version would be fine indeed.

bhalevy commented on replica/table.cc:
```diff
@@ -933,7 +940,7 @@ compaction_group& tablet_storage_group_manager::compaction_group_for_sstable(con
     }
 
-    auto& sg = storage_groups()[first_id];
+    auto& sg = storage_groups().at(first_id);
```
ditto, check for tablet existence and abort with internal_error if it's missing instead of terminating due to throwing in noexcept context.

bhalevy commented on replica/tablets.cc:
```diff
@@ -361,15 +364,13 @@ class tablet_sstable_set : public sstables::sstable_set_impl {
-        _sstable_sets.reserve(sgm.storage_groups().size());
-        for (const auto& sg : sgm.storage_groups()) {
+        for (const auto& [id, sg] : sgm.storage_groups()) {
             if (sg) {
```
else?  can this really happen or is it an internal error with the flat_hash_map?

raphaelsc replied:
It can go away later. I don't think we need unique_ptr anymore. sg will never be nullptr here.

bhalevy replied:
possibly so, when splitting we create new storage_groups anyhow.

bhalevy commented on replica/tablets.cc:
```diff
@@ -461,10 +468,8 @@ future<std::optional<tablet_transition_stage>> read_tablet_transition_stage(cql3
-    for (auto i = candidate_start; i <= candidate_end; i++) {
-        if (const auto& set = _sstable_sets[i]) {
+    for (const auto& i : subrange(pr)) {
+        if (const auto& set = _sstable_sets.at(i)) {
```
else?  isn't it internal error?

raphaelsc replied:
sg will never be nullptr here. we can stop using unique_ptr and this if will go away. we just need to switch to get_sstable_set(i) that finds i first and abort if not found.

bhalevy commented on replica/tablets.cc:
```diff
@@ -556,11 +559,11 @@ class tablet_incremental_selector : public sstables::incremental_selector_impl {
             auto idx = _tset.group_of(token);
             if (!token.is_maximum()) {
-                _cur_set = _tset._sstable_sets[idx];
+                _cur_set = _tset._sstable_sets.at(idx);
```
this creates a regression.  we don't require the set to exist.  there's code a few lines below to handle `if (!_cur_set)` and return an empty list, plus a hint about the lowest_next_position.

raphaelsc replied:
why? we won't go through inexistent sets anymore.

raphaelsc replied:
this was only needed when you emplaced non-engaged sets.

bhalevy replied:
But select can be called with any token

raphaelsc replied:
with tokens that are owned by existing tablets. what can happen is that we won't find the sstable_set, and that will be handled with your proposed find_sstable_set that aborts if id is not found, but a sstable set at an id X won't be empty as before with the sparse array. we don't have a sparse array anymore.

avikivity replied:
select can be called with any token, and storage_proxy will consult effective_replication_map to find the node that owns the token, and fencing will prevent an oudated coordinator from reaching this area.

bhalevy commented on replica/tablets.cc:
```diff
@@ -629,7 +630,7 @@ tablet_sstable_set::create_single_key_sstable_reader(
     // The singular partition_range start bound must be engaged.
     auto idx = group_of(pr.start()->value().token());
-    const auto& set = _sstable_sets[idx];
+    const auto& set = _sstable_sets.at(idx);
```
use `find` instead so we can handle the non-existing set below.

bhalevy reviewed (changes requested):
see comments in line

bhalevy commented on replica/table.cc:
```diff
@@ -882,7 +888,7 @@ storage_group* table::storage_group_for_token(dht::token token) const noexcept {
 compaction_group& tablet_storage_group_manager::compaction_group_for_token(dht::token token) const noexcept {
     auto [idx, range_side] = storage_group_of(token);
-    auto& sg = *storage_groups()[idx];
+    auto& sg = *storage_groups().at(idx);
```
sounds good

raphaelsc commented:
v3: 
employ sanity checks when finding storage group and sstable set

raphaelsc commented:
@bhalevy please re-review.

bhalevy commented on replica/table.cc:
```diff
@@ -549,8 +549,17 @@ void table::enable_off_strategy_trigger() {
 
 storage_group_manager::~storage_group_manager() = default;
 
+storage_group* storage_group_manager::storage_group_for_id(const schema_ptr& s, size_t i) const {
```
I think we can just modify `storage_group_of` to return `std::pair<storage_group*, locator::tablet_range_side>`

bhalevy commented on replica/tablets.cc:
```diff
@@ -461,10 +477,8 @@ future<std::optional<tablet_transition_stage>> read_tablet_transition_stage(cql3
-    for (auto i = candidate_start; i <= candidate_end; i++) {
-        if (const auto& set = _sstable_sets[i]) {
+    for (const auto& i : subrange(pr)) {
+        if (const auto& set = find_sstable_set(i)) {
```
if `find_sstable_set(i)` always returns an engaged lw_shared_ptr we don't need to the condition, do we?

raphaelsc replied:
yes, I am planning to remove it later to reduce changes in the patch.

bhalevy commented on replica/tablets.cc:
```diff
@@ -556,11 +568,11 @@ class tablet_incremental_selector : public sstables::incremental_selector_impl {
             auto idx = _tset.group_of(token);
             if (!token.is_maximum()) {
-                _cur_set = _tset._sstable_sets[idx];
+                _cur_set = _tset.find_sstable_set(idx);
```
I hope that the assumption that the search token must be covered by an allocated tablet holds water.

bhalevy commented on replica/tablets.cc:
```diff
@@ -629,7 +639,7 @@ tablet_sstable_set::create_single_key_sstable_reader(
     // The singular partition_range start bound must be engaged.
     auto idx = group_of(pr.start()->value().token());
-    const auto& set = _sstable_sets[idx];
+    const auto& set = find_sstable_set(idx);
```
if `find_sstable_set` always returns an engaged set, the condition below is moot and can be removed.

bhalevy commented on replica/compaction_group.hh:
```diff
@@ -243,12 +244,13 @@ public:
+    storage_group_map& storage_groups() noexcept {
         return _storage_groups;
     }
+    storage_group* storage_group_for_id(const schema_ptr&, size_t i) const;
```
Please document that it always returns a non-null pointer.
This FIXME comment for the table::storage_group_for_id entry point applies here too.

bhalevy replied:
Also, can be virtual, with a trivial implementation for the single group manager, but I don't know if it's worth it.

raphaelsc commented:
v4:
- Added FIXME for changing storage_group_for_id signature later.
- Removed needed handling for unengaged sstable sets

Deexie commented on replica/table.cc:
```diff
@@ -820,7 +838,7 @@ sstables::compaction_type_options::split tablet_storage_group_manager::split_com
     sstables::compaction_type_options::split opt = split_compaction_options();
 
-    for (auto& storage_group : storage_groups()) {
+    for (auto& storage_group : storage_groups() | boost::adaptors::map_values) {
```
Please note, that with dynamic groups allocation we would need to get back to iterating over all possible (`tablet_count()`) storage groups here. Otherwise, if the loop preempts, iterators may get invalidated. (I think that's the easiest way to prevent it)

Other loops over storage groups don't preempt.

bhalevy replied:
I think we should allocate and deallocate storage_groups only via update_effective_replication_map, and so we can use a rwlock to protect storage_groups() while it's scanned.

Deexie replied:
I see. Currently storage groups are deallocated in tablet_cleanup (cg needs to be stopped there and we can't keep it after it's stopped), but it doesn't change much I think

avikivity replied:
This should be called with effective_replication_map_ptr held, so no changes can happen during iteration. Please verify that.

Deexie replied:
Maybe I'm missing something but I do not see where it could be held

avikivity replied:
It must be held. How can we split storage groups if someone else can move a new one in or an old one out?

avikivity replied:
Practically any call here should be with the effective replication map locked, one way or another. It's just too fragile otherwise.

raphaelsc replied:
if erm is taken after pending updates reached replica, those updates could still be applied. we're thinking about using a rwlock to synchronize changes to erm and scanning of storage groups, that reflect current erm.

for example:

1) split and migration start concurrently
2) migration advances to allow_write...
3) split loop starts and take erm, and preempt
3) migration update reaches replica and causes replica table to allocate a new storage group

maybe preemptive loops over storage groups should hold erm in addition to the rwlock.

raphaelsc replied:
actually, I think we shouldn't hold erm during split_all_groups(). otherwise, we lose the ability to split groups concurrently to migration. rwlock should be enough, right?. when allocating / deallocating / splitting groups, we'll synchronize with any preemptive loops over groups through rwlock.

raphaelsc replied:
and splitting groups can take long, blocking topology changes for a significant amount of time. let's say a shard has 1T of data, with compaction at 30M/s/shard. let's assume 100G intersecting (100 tablets) with split point. That's almost 1 hour.

bhalevy replied:
Currently the preparation work for splitting is not not changing topology.
Only when all tablets are split, we apply the topology change and "promote" the split compaction groups to be first-class storage groups.  The is done via update_effective_replication_map that must be atomic with respect the to storage group manager.  Therefore tablet mugration sode effects of adding or removing tablet ownership should be applied using the same path and the storage group manager should update its storage groups based on the delta between the current erm and the new erm.

bhalevy replied:
Cc @Deexie ☝️

raphaelsc commented:
@bhalevy please re-review.

bhalevy reviewed (approved):
LGTM

avikivity commented on replica/table.cc:
```diff
@@ -549,8 +549,17 @@ void table::enable_off_strategy_trigger() {
+storage_group* storage_group_manager::storage_group_for_id(const schema_ptr& s, size_t i) const {
+    auto it = storage_groups().find(i);
+    if (it == storage_groups().end()) [[unlikely]] {
+        on_fatal_internal_error(tlogger, format("Storage wasn't found for tablet {} of table {}.{}", i, s->ks_name(), s->cf_name()));
```
Why fatal? We can continue with an exception in case we have such a bug.

raphaelsc replied:
agreed. I was thinking about it yesterday, plain exception or on_internal_error is enough?

avikivity commented on replica/table.cc:
```diff
@@ -650,7 +668,7 @@ class tablet_storage_group_manager final : public storage_group_manager {
             }
             // FIXME: don't allocate compaction groups for tablets that aren't present in this shard.
```
remove?

avikivity replied:
or do you still allocate

raphaelsc replied:
I'll leave it for @Deexie, in her PR. we're still allocating all tablet ids, as before.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7856/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7856/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7856/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/831/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8865/)
#### Build Details:
- Duration: 4 hr 35 min
- Builder: spider1.cloudius-systems.com

raphaelsc commented:
@avikivity please review again.
